File: Java/Swiat.java
import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.*;

public class Swiat {
    private int x;
    private int y;
    private int tura;
    private Organizm[][] plansza;
    private boolean czyCzlowiekZyje;
    private boolean pauza;
    private ArrayList<Organizm> organizmy;
    private Czlowiek czlowiekWsk;
    private SwiatGUI swiatGUI;

    public Swiat(SwiatGUI swiatGUI){
        this.x= 0;
        this.y = 0;
        tura = 0;
        czyCzlowiekZyje = true;
        pauza = true;
        Organizm[] organizmy = null;
        this.swiatGUI = swiatGUI;
    }

    public Swiat(int sizeX, int sizeY, SwiatGUI swiatGUI) {
        this.x= sizeX;
        this.y= sizeY;
        tura = 0;
        czyCzlowiekZyje = true;
        pauza = true;
        plansza = new Organizm[sizeY][sizeX];
        for (int i = 0; i < sizeY; i++) {
            for (int j = 0; j < sizeX; j++) {
                plansza[i][j] = null;
            }
        }
        organizmy = new ArrayList<>();
        this.swiatGUI = swiatGUI;
    }

    protected void stworzSwiat(double wypelnienie){
        this.czlowiekWsk = (Czlowiek) DodajOrganizm.nowyOrganizm(this, 0, 0, Organizm.TypOrganizmu.CZLOWIEK);
        DodajOrganizm(czlowiekWsk);
        Random random = new Random();
        int x,y; wypelnienie=wypelnienie*10*2;
        for (int i = 0; i < wypelnienie; i++)
        {
            x = random.nextInt(this.x);
            y = random.nextInt(this.y);
            Organizm.TypOrganizmu[] table = {
                    Organizm.TypOrganizmu.ANTYLOPA, Organizm.TypOrganizmu.WILCZE_JAGODY, Organizm.TypOrganizmu.BARSZCZ_SOSNOWSKIEGO,
                    Organizm.TypOrganizmu.LIS, Organizm.TypOrganizmu.TRAWA, Organizm.TypOrganizmu.GUARANA, Organizm.TypOrganizmu.OWCA,
                    Organizm.TypOrganizmu.MLECZ, Organizm.TypOrganizmu.ZOLW, Organizm.TypOrganizmu.WILK
            };
            if (this.plansza[y][x]==null)
                DodajOrganizm(DodajOrganizm.nowyOrganizm(this, x, y, table[random.nextInt(table.length)]));
        }
    }
    void zarzadzanieRuchem(int fromX, int fromY, int toX, int toY)
    {
        System.out.println(fromX+" x "+fromY+" y x1 "+toX+" y1 "+toY);
        if (this.plansza[toY][toX]==null)
        {
            System.out.println("Rusza się na puste");
            this.zamienMiejscami(fromX, fromY, toX, toY);
            return;
        };
        if (this.plansza[fromY][fromX].getZnak() == this.plansza[toY][toX].getZnak() &&
                this.plansza[fromY][fromX] != this.plansza[toY][toX])			//rozmnazanie
        {
            System.out.println("Rozmnaza sie");
            if (!this.plansza[toY][toX].getRozmnozylSie() && !this.plansza[fromY][fromX].getRozmnozylSie() && this.plansza[fromY][fromX].rozmnoz(toX, toY)) {
                this.plansza[fromY][fromX].setRozmnozylSie(true); this.plansza[toY][toX].setRozmnozylSie(true);
        }
        }
	    else
        {
            if (this.plansza[fromY][fromX] != this.plansza[toY][toX]) {
            if (!this.plansza[fromY][fromX].kolizja(this.plansza[toY][toX])) return;
            if (!this.plansza[toY][toX].kolizja(this.plansza[fromY][fromX])) return;
            //walka
            if (this.plansza[fromY][fromX].getSila() >= this.plansza[toY][toX].getSila())
            {
                System.out.println("Zjada ktos kogos");
                this.UsunOrganizm((this.plansza[toY][toX]));
                this.zamienMiejscami(fromX, fromY, toX, toY);
            }
			else
            {
                System.out.println("Umiera");
                this.UsunOrganizm((this.plansza[fromY][fromX]));
            }
        }
        };
    }
    void barszczSmierci(int x, int y) {
        if (this.plansza[y][x] instanceof Zwierze)
            this.UsunOrganizm(plansza[y][x]);
    }
    void zamienMiejscami(int ad1X, int ad1Y, int ad2X, int ad2Y)
    {

        if (this.plansza[ad1Y][ad1X]!=null)
        {
            this.plansza[ad1Y][ad1X].setX(ad2X);
            this.plansza[ad1Y][ad1X].setY(ad2Y);
        }
        if (this.plansza[ad2Y][ad2X] != null)
        {
            this.plansza[ad2Y][ad2X].setY(ad1Y);
            this.plansza[ad2Y][ad2X].setX(ad1X);
        }
        Organizm temp = plansza[ad1Y][ad1X];
        plansza[ad1Y][ad1X] = plansza[ad2Y][ad2X];
        plansza[ad2Y][ad2X] = temp;
    }
    boolean pustePole(int x, int y){
        if(plansza[y][x]!=null)
            return false;
        else
            return true;
    }
    int getX(){
        return x;
    }
    int getY(){
        return y;
    }

    void zapisz(String nameOfFile){
        try {
            nameOfFile += ".txt";
            File file = new File(nameOfFile);
            file.createNewFile();

            PrintWriter pw = new PrintWriter(file);
            pw.print(x + " ");
            pw.print(y + " ");
            pw.print(tura + " ");
            pw.print(czyCzlowiekZyje + " ");
            for (int i = 0; i < organizmy.size(); i++) {
                pw.print(organizmy.get(i).getZnak() + " ");
                pw.print(organizmy.get(i).getX() + " ");
                pw.print(organizmy.get(i).getY() + " ");
                pw.print(organizmy.get(i).getSila() + " ");
                pw.print(organizmy.get(i).getWiek() + " ");
                pw.print(organizmy.get(i).getRozmnozylSie());
                if (organizmy.get(i).getZnak() == Organizm.TypOrganizmu.CZLOWIEK) {
                    pw.print(" " + czlowiekWsk.getSuperMocCount() + " ");
                    pw.print(czlowiekWsk.getSuperMocCd() + " ");
                }
                pw.println();
            }
            pw.close();
        } catch (IOException e) {
            System.out.println("Error: " + e);
        }
    }
    public static Swiat wczytaj(String nameOfFile){
        try {
            nameOfFile += ".txt";
            File file = new File(nameOfFile);

            Scanner scanner = new Scanner(file);
            String line = scanner.nextLine();
            String[] properties = line.split(" ");
            int sizeX = Integer.parseInt(properties[0]);
            int sizeY = Integer.parseInt(properties[1]);
            Swiat tmpSwiat = new Swiat(sizeX, sizeY, null);
            int numerTury = Integer.parseInt(properties[2]);
            tmpSwiat.tura = numerTury;
            boolean czyCzlowiekZyje = Boolean.parseBoolean(properties[3]);
            tmpSwiat.czyCzlowiekZyje = czyCzlowiekZyje;
            tmpSwiat.czlowiekWsk = null;

            while (scanner.hasNextLine()) {
                line = scanner.nextLine();
                properties = line.split(" ");
                Organizm.TypOrganizmu typOrganizmu = Organizm.TypOrganizmu.valueOf(properties[0]);
                int x = Integer.parseInt(properties[1]);
                int y = Integer.parseInt(properties[2]);

                Organizm tmpOrganizm = DodajOrganizm.nowyOrganizm(tmpSwiat,x,y,typOrganizmu);
                int sila = Integer.parseInt(properties[3]);
                tmpOrganizm.setSila(sila);
                int turaUrodzenia = Integer.parseInt(properties[4]);
                tmpOrganizm.setInicjatywaWiek(tmpOrganizm.getInicjatywa(),turaUrodzenia);
                boolean czyRozmnozyl = Boolean.parseBoolean(properties[5]);
                tmpOrganizm.setRozmnozylSie(czyRozmnozyl);

                if (typOrganizmu == Organizm.TypOrganizmu.CZLOWIEK) {
                    tmpSwiat.czlowiekWsk = (Czlowiek) tmpOrganizm;
                    int czasTrwania = Integer.parseInt(properties[6]);
                    tmpSwiat.czlowiekWsk.setSuperMocCount(czasTrwania);
                    int cooldown = Integer.parseInt(properties[7]);
                    tmpSwiat.czlowiekWsk.setSuperMocCd(cooldown);
                }
                tmpSwiat.DodajOrganizm(tmpOrganizm);
            }
            scanner.close();
            return tmpSwiat;
        } catch (
                IOException e) {
            System.out.println("Error: " + e);
        }
        return null;
    }
    public void DodajOrganizm(Organizm organizm) {
        organizmy.add(organizm);
        System.out.println(organizm.getX()+" "+organizm.getY());
        plansza[organizm.getY()][organizm.getX()] = organizm;
    }
    void nowaTura(){
        tura++;
        Dziennik.dodajWpis("\nTURA " + tura);
        System.out.println(tura);
        System.out.println(organizmy.size() + "\n");
        SortujOrganizmy();
        for (int i = 0; i < organizmy.size(); i++) {
            if (organizmy.get(i).getWiek() != tura && organizmy.get(i).getCzyUmarl() == false) {
                organizmy.get(i).wykonajRuch();
                System.out.println(organizmy.get(i).getZnak());
            }
        }
        for (int i = 0; i < organizmy.size(); i++) {
            if (organizmy.get(i).getCzyUmarl() == true) {
                organizmy.remove(i);
                i--;
            }
        }
        for (int i = 0; i < organizmy.size(); i++) {
            organizmy.get(i).setRozmnozylSie(false);
        }
    }

    private void SortujOrganizmy() {
        Collections.sort(organizmy, new Comparator<Organizm>() {
            @Override
            public int compare(Organizm o1, Organizm o2) {
                if (o1.getInicjatywa() != o2.getInicjatywa())
                    return Integer.valueOf(o2.getInicjatywa()).compareTo(o1.getInicjatywa());
                else
                    return Integer.valueOf(o1.getWiek()).compareTo(o2.getWiek());
            }
        });
    }
    public void UsunOrganizm(Organizm organizm) {
        plansza[organizm.getY()][organizm.getX()] = null;
        organizm.setCzyUmarl(true);
        if (organizm.getZnak() == Organizm.TypOrganizmu.CZLOWIEK) {
            czyCzlowiekZyje = false;
            czlowiekWsk = null;
        }
    }
    public SwiatGUI getSwiatGUI() {
        return swiatGUI;
    }

    public void setSwiatGUI(SwiatGUI swiatGUI) {
        this.swiatGUI = swiatGUI;
    }

    public boolean isPauza() {
        return pauza;
    }

    public void setPauza(boolean pauza) {
        this.pauza = pauza;
    }

    public Czlowiek getCzlowiek() {
        return czlowiekWsk;
    }
    public Organizm[][] getPlansza() {
        return plansza;
    }

    public void setCzlowiek(Czlowiek czlowiek) {
        this.czlowiekWsk = czlowiek;
    }
    public void setCzyCzlowiekZyje(boolean czyCzlowiekZyje) {
        this.czyCzlowiekZyje = czyCzlowiekZyje;
    }
    public boolean getCzyCzlowiekZyje() {
        return czyCzlowiekZyje;
    }
    boolean wGranicach(int x, int y)
    {
        if (x < this.x && x >= 0 && y >= 0 && y < this.y)
        {
            return true;
        }
        return false;
    }
}

File: C++/Swiat.cpp
#include "Swiat.h"
#include "Organizm.h"
#include <iostream>
#include <time.h>

#include "WilczeJagody.h"
#include "Barszcz.h"
#include "Antylopa.h"
#include "Lis.h"
#include "Trawa.h"
#include "Guarana.h"
#include "Czlowiek.h"
#include "Owca.h"
#include "Mlecz.h"
#include "Zolw.h"
#include "Wilk.h"

#include "Defines.h"

#include <fstream>

Swiat::Swiat(int x, int y)
{
	this->rozmiarSwiata.x = x;
	this->rozmiarSwiata.y = y;
	this->tura = 0;
	this->Organizmy = new Organizm * [x * y];
	for (int i = 0; i < x * y; i++)
	{
		this->Organizmy[i] = NULL;
	}
}

Swiat::~Swiat()
{
	delete[] Organizmy;
}

void Swiat::start()
{
	int liczba;
	cout << "1. Nowa gra \n2. Wczytaj" << endl;
	cin >> liczba;
	if (liczba == 1) this->stworzSwiat();
	if (liczba == 2) this->wczytaj();
	if (liczba > 2)
	{
		exit(-1);
	}
	while (true) this->nowaTura();
}

void Swiat::stworzSwiat()
{
	this->czlowiekWsk = (Czlowiek*)this->nowyOrganizm(0, 0, 'C');
	int x, y;
	for (int i = 0; i < START_OBJ_NUMBER; i++)
	{
		x = rand() % rozmiarSwiata.x;
		y = rand() % rozmiarSwiata.y;
		typOrganizmu table[] = { ANTYLOPA, WILCZEJAGODY, BARSZCZ, LIS, TRAWA, GUARANA, OWCA, MLECZ, ZOLW, WILK };
		typOrganizmu random = table[rand() % NUMBER_OF_DIFFRENT_ORGANISMS];
		if (this->Organizmy[x * this->rozmiarSwiata.y + y] == NULL)
			this->nowyOrganizm(x, y, random);
	}
	this->rysujSwiat();
}

void Swiat::nowaTura()
{
	this->tura++;
	cout << "\nTurn " << tura << endl;
	for (int inicjatywa = MAX_INIT - 1; inicjatywa >= 0; inicjatywa--)
	{
		for (int ageReverse = tablicaInicjatywy[inicjatywa].size() - 1; ageReverse >= 0; ageReverse--)
		{
			//cout << this->tablicaInicjatywy[inicjatywa][ageReverse]->getWiek() << " " << tablicaInicjatywy[inicjatywa][ageReverse]->getZnak() << " " << tablicaInicjatywy[inicjatywa][ageReverse]->getInicjatywa() << endl;
			this->tablicaInicjatywy[inicjatywa][ageReverse]->wykonajRuch();
		}
		for (int ageReverse = tablicaInicjatywy[inicjatywa].size() - 1; ageReverse >= 0; ageReverse--) {
			this->tablicaInicjatywy[inicjatywa][ageReverse]->starzejSie();
			this->tablicaInicjatywy[inicjatywa][ageReverse]->setRozmnozylSie(false);
		}
	}
	//getchar();
	this->rysujSwiat();
}

void Swiat::rysujSwiat()
{
	system("cls");
	cout << "Czlowiek sila: " << this->czlowiekWsk->getSila() << "\n";
	for (int i = 0; i <= this->rozmiarSwiata.y + 1; i++)cout << "#";
	cout << "\n#";
	for (int x = 0; x < this->rozmiarSwiata.x; x++)
	{
		for (int y = 0; y < this->rozmiarSwiata.y; y++)
		{
			if (this->Organizmy[x * this->rozmiarSwiata.y + y] == NULL)
			{
				cout << " ";
			}
			else
			{
				cout << this->Organizmy[x * this->rozmiarSwiata.y + y]->getZnak();
			}
		}
		cout << "#\n#";
	}
	for (int i = 0; i < this->rozmiarSwiata.y + 1; i++)cout << "#";
	this->dziennik.aktualizujDziennik();
}

void Swiat::zarzadzanieRuchem(int fromX, int fromY, int toX, int toY)
{
	int from = fromX * this->rozmiarSwiata.y + fromY, to = toX * this->rozmiarSwiata.y + toY;
	if (this->Organizmy[from] == czlowiekWsk)
		this->dziennik.czlowiekSuperMoc(czlowiekWsk->getSuperMocCount());
	if (this->Organizmy[to] == NULL)
	{
		this->zamienMiejscami(fromX, fromY, toX, toY);
		return;
	};
	if (this->Organizmy[to]->getZnak() == this->Organizmy[from]->getZnak() && this->Organizmy[to] != this->Organizmy[from])			//rozmnazanie
	{
		if (!this->Organizmy[to]->getRozmnozylSie() && !this->Organizmy[from]->getRozmnozylSie() && this->Organizmy[from]->rozmnoz(toX, toY)) {
			this->Organizmy[from]->setRozmnozylSie(true); this->Organizmy[to]->setRozmnozylSie(true);
			this->dziennik.opisRozmnazania(this->Organizmy[from]);
		}
	}
	else
	{
		if (!this->Organizmy[from]->kolizja(this->Organizmy[to]))return;
		if (!this->Organizmy[to]->kolizja(this->Organizmy[from]))return;
		//walka
		if (this->Organizmy[from]->getSila() >= this->Organizmy[to]->getSila())
		{
			this->dziennik.opisWalki(this->Organizmy[from], this->Organizmy[to]);
			this->usunOrganizm(*(this->Organizmy[to]));
		}
		else
		{
			this->dziennik.opisWalki(this->Organizmy[to], this->Organizmy[from]);
			this->usunOrganizm(*(this->Organizmy[from]));
		}
		this->zamienMiejscami(fromX, fromY, toX, toY);
	};
}

void Swiat::usunOrganizm(Organizm& organizm)
{
	int inicjatywa = organizm.getInicjatywa();
	for (vector <Organizm*>::iterator i = this->tablicaInicjatywy[inicjatywa].begin(); i < this->tablicaInicjatywy[inicjatywa].end(); i++)
	{
		if (*i == &organizm)
		{
			this->tablicaInicjatywy[inicjatywa].erase(i);
			break;
		}
	}
	this->Organizmy[organizm.getX() * this->rozmiarSwiata.y + organizm.getY()] = NULL;
	delete& organizm;
}

void Swiat::barszczSmierci(int ktory) {
	if (((this->Organizmy[ktory])->getZnak()) != 'b')
		this->usunOrganizm(*(this->Organizmy[ktory]));
}

bool Swiat::pustePole(int x, int y)
{
	if (!this->wGranicach(x, y))return false;
	if (this->Organizmy[x * this->rozmiarSwiata.y + y] == NULL)return true;
	return false;
}

bool Swiat::wGranicach(int x, int y)
{
	if (x < this->rozmiarSwiata.x && x >= 0 && y >= 0 && y < this->rozmiarSwiata.y)
	{
		return true;
	}
	return false;
}

int Swiat::getX() {
	return this->rozmiarSwiata.x;
}

int Swiat::getY() {
	return this->rozmiarSwiata.y;
}

void Swiat::zamienMiejscami(int ad1X, int ad1Y, int ad2X, int ad2Y)
{
	int pierwszy = ad1X * this->rozmiarSwiata.y + ad1Y,
		drugi = ad2X * this->rozmiarSwiata.y + ad2Y;

	if (this->Organizmy[pierwszy] != NULL)
	{
		this->Organizmy[pierwszy]->setX(ad2X);
		this->Organizmy[pierwszy]->setY(ad2Y);
	}
	if (this->Organizmy[drugi] != NULL)
	{
		this->Organizmy[drugi]->setY(ad1Y);
		this->Organizmy[drugi]->setX(ad1X);
	}
	swap(this->Organizmy[pierwszy], this->Organizmy[drugi]);
	return;
}

void Swiat::zapisz()
{
	fstream plik;
	cout << "Podan nazwe pliku:" << endl;
	string nazwa;
	cin >> nazwa;
	plik.open(nazwa, ios::out | ios::trunc);
	plik << this->tura << endl;
	plik << this->rozmiarSwiata.x << " " << this->rozmiarSwiata.y << endl;
	for (int i = 0; i < this->rozmiarSwiata.x * this->rozmiarSwiata.y; i++)
	{
		if (this->Organizmy[i] != NULL)
		{
			plik << this->Organizmy[i]->getZnak() << " ";
			plik << this->Organizmy[i]->getWiek() << " ";
			plik << this->Organizmy[i]->getSila() << " ";
			plik << this->Organizmy[i]->getInicjatywa() << " ";
			plik << this->Organizmy[i]->getX() << " ";
			plik << this->Organizmy[i]->getY() << " ";
			if (this->Organizmy[i] == this->czlowiekWsk)plik << this->czlowiekWsk->getSuperMocCount() << " " << this->czlowiekWsk->getSuperMocCd() << " ";
			plik << endl;
		}
	}
	plik.close();
}

void Swiat::wczytaj()
{
	fstream plik;
	cout << "Podaj nazwe pliku" << endl;
	string nazwa;
	cin >> nazwa;
	plik.open(nazwa, ios::in);
	if (!plik.good())return;
	plik >> this->tura;
	plik >> this->rozmiarSwiata.x;
	plik >> this->rozmiarSwiata.y;
	char znak;
	int wiek, sila, inicjatywa, Posx, Posy, cd, count;
	while (true) {
		if (!(plik >> znak))
			break;
		plik >> wiek >> sila >> inicjatywa >> Posx >> Posy;
		if (znak != 'C')
		{
			Organizm* ptr = NULL;
			ptr = this->nowyOrganizm(Posx, Posy, znak);
			ptr->setInicjatywaWiek(inicjatywa, wiek);
			ptr->setSila(sila);
		}
		else
		{
			plik >> count >> cd;
			this->czlowiekWsk = (Czlowiek*)this->nowyOrganizm(Posx, Posy, 'C');
			this->czlowiekWsk->setInicjatywaWiek(inicjatywa, wiek);
			this->czlowiekWsk->setSila(sila);
			this->czlowiekWsk->setSuperMocCd(cd);
			this->czlowiekWsk->setSuperMocCount(count);
		}
	}
	plik.close();
	this->rysujSwiat();
}

void Swiat::aktualizujTabliceInicjatywy(Organizm* pierwszy, int drugiInicjatywa, int drugiWiek)
{
	if (this->tablicaInicjatywy[pierwszy->getInicjatywa()].size() == 1)
	{
		this->tablicaInicjatywy[pierwszy->getInicjatywa()].clear();
	}
	else
	{
		for (vector <Organizm*>::iterator iter = this->tablicaInicjatywy[pierwszy->getInicjatywa()].begin(); iter < this->tablicaInicjatywy[pierwszy->getInicjatywa()].end(); iter++)
		{
			if (*iter == pierwszy)
			{
				this->tablicaInicjatywy[pierwszy->getInicjatywa()].erase(iter);
				break;
			}
		}
	}
	if (!this->tablicaInicjatywy[drugiInicjatywa].empty())
	{
		for (vector <Organizm*>::iterator iter = this->tablicaInicjatywy[drugiInicjatywa].begin(); iter < this->tablicaInicjatywy[drugiInicjatywa].end(); iter++)
		{
			if ((*iter)->getWiek() >= drugiWiek)
			{
				this->tablicaInicjatywy[drugiInicjatywa].insert(iter, pierwszy);
				break;
			}
			else if (iter == this->tablicaInicjatywy[drugiInicjatywa].end() - 1)
			{
				this->tablicaInicjatywy[drugiInicjatywa].push_back(pierwszy);
			}
		}
	}
	else
	{
		this->tablicaInicjatywy[drugiInicjatywa].push_back(pierwszy);
	}
	return;
}

Organizm* Swiat::nowyOrganizm(int x, int y, char znak)
{
	int pozycja = x * (this->rozmiarSwiata.y) + y;
	if (this->Organizmy[pozycja] != NULL)
	{
		return NULL;
	}
	else
	{
		Organizm* ptr = NULL;
		switch (znak)
		{
		case 'A':
		{
			ptr = new Antylopa(*this, x, y);
			break;
		}
		case 'j':
		{
			ptr = new WilczeJagody(*this, x, y);
			break;
		}
		case 'b':
		{
			ptr = new Barszcz(*this, x, y);
			break;
		}
		case 'L':
		{
			ptr = new Lis(*this, x, y);
			break;
		}
		case 't':
		{
			ptr = new Trawa(*this, x, y);
			break;
		}
		case 'g':
		{
			ptr = new Guarana(*this, x, y);
			break;
		}
		case 'C':
		{
			ptr = new Czlowiek(*this, x, y);
			break;
		}
		case 'O':
		{
			ptr = new Owca(*this, x, y);
			break;
		}
		case 'm':
		{
			ptr = new Mlecz(*this, x, y);
			break;
		}
		case 'Z':
		{
			ptr = new Zolw(*this, x, y);
			break;
		}
		case 'W':
		{
			ptr = new Wilk(*this, x, y);
			break;
		}
		default:
		{
			throw "Unable to read znak";
		}
		}

		this->Organizmy[pozycja] = ptr;
		this->tablicaInicjatywy[ptr->getInicjatywa()].push_back(ptr);
		return ptr;
	}
};